File: src/accounting/patterns/MultiplyByRatePostingRule.java
package accounting.patterns;

import java.math.BigDecimal;
import java.util.Currency;

class MultiplyByRatePostingRule extends PostingRule {
    public MultiplyByRatePostingRule(EntryType type) {
        super(type);
    }

    protected MonetaryAmount calculateAmount(AccountingEvent evt) {

        Usage usageEvent = (Usage) evt;
        return new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(usageEvent.getAmount().getAmount() * usageEvent.getRate()));
    }
}

File: src/main/typescript/accounting-pattern/MultiplyByRatePostingRule.ts
import type { EntryType } from "./EntryType";
import { MonetaryAmount } from "./MonetaryAmount";
import { PostingRule } from "./PostingRule";
import type { Quantity } from "./Quantity";

/** Charges the agreement's rate for every unit of quantity. */
export class MultiplyByRatePostingRule extends PostingRule {
    constructor(type: EntryType) {
        super(type);
    }

    protected calculateAmount(quantity: Quantity, rate: number): MonetaryAmount {
        return new MonetaryAmount("USD", quantity.getValue() * rate);
    }
}
